File: Java8-JavaFX/src/sevenguis/crud/CRUD.java
package sevenguis.crud;

import javafx.application.Application;
import javafx.beans.binding.IntegerExpression;
import javafx.beans.binding.StringExpression;
import javafx.collections.FXCollections;
import javafx.collections.ListChangeListener;
import javafx.collections.ObservableList;
import javafx.collections.transformation.FilteredList;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.HBox;
import javafx.stage.Stage;

import java.util.ArrayList;
import java.util.List;

public class CRUD extends Application {

    public void start(Stage stage) {
        TextField prefix = new TextField();
        prefix.setPrefWidth(60);
        TextField name = new TextField();
        name.setPrefWidth(100);
        TextField surname = new TextField();
        surname.setPrefWidth(100);
        Button create = new Button("Create");
        Button update = new Button("Update");
        Button delete = new Button("Delete");
        update.setDisable(true);
        delete.setDisable(true);
        ListView<String> entries = new ListView<>();
        entries.getSelectionModel().setSelectionMode(SelectionMode.SINGLE);

        List<String> externDb = new ArrayList<>();
        externDb.add("Emil, Hans");
        externDb.add("Musterman, Max");
        externDb.add("Tisch, Roman");
        ObservableList<String> db = FXCollections.observableArrayList(externDb);
        db.addListener((ListChangeListener<String>) c -> {
            while (c.next()) {
                if (c.wasReplaced()) externDb.set(c.getFrom(), c.getAddedSubList().get(0));
                else {
                    if (c.wasAdded()) externDb.add(c.getAddedSubList().get(0));
                    if (c.wasRemoved()) externDb.remove(c.getFrom());
                }
            }
        });
        FilteredList<String> dbView = db.filtered(item -> true);
        entries.setItems(dbView);

        StringExpression fullname = surname.textProperty().concat(", ").concat(name.textProperty());
        IntegerExpression selectedIndex = entries.getSelectionModel().selectedIndexProperty();
        prefix.textProperty().addListener((v, o, n) -> dbView.setPredicate(item -> item.startsWith(n)));
        create.setOnAction(e -> db.add(fullname.get()));
        delete.setOnAction(e -> db.remove(dbView.getSourceIndex(selectedIndex.get())));
        update.setOnAction(e -> db.set(dbView.getSourceIndex(selectedIndex.get()), fullname.get()));
        delete.disableProperty().bind(selectedIndex.isEqualTo(-1));
        update.disableProperty().bind(selectedIndex.isEqualTo(-1));

        BorderPane root = new BorderPane();
        root.setPrefSize(400, 400);
        root.setPadding(new Insets(10));
        HBox top = new HBox(10, new Label("Filter prefix: "), prefix);
        top.setPadding(new Insets(0, 0, 10, 0));
        top.setAlignment(Pos.BASELINE_LEFT);
        root.setTop(top);
        root.setCenter(entries);
        GridPane right = new GridPane();
        right.setHgap(10);
        right.setVgap(10);
        right.setPadding(new Insets(0, 0, 0, 10));
        right.addRow(0, new Label("Name: "), name);
        right.addRow(1, new Label("Surname: "), surname);
        root.setRight(right);
        HBox bottom = new HBox(10, create, update, delete);
        bottom.setPadding(new Insets(10, 0, 0, 0));
        root.setBottom(bottom);

        stage.setScene(new Scene(root));
        stage.setTitle("CRUD");
        stage.show();
    }

    public static void main(String[] args) {
        launch(args);
    }
}

File: Scala-ScalaFX/src/main/scala/sevenguis/crud/CRUD.scala
package sevenguis.crud

import scalafx.application.JFXApp
import scalafx.application.JFXApp.PrimaryStage
import scalafx.scene.Scene
import scalafx.scene.control._
import scalafx.scene.layout.{GridPane, BorderPane, HBox}
import scalafx.geometry.{Pos, Insets}
import scalafx.event.ActionEvent
import scalafx.Includes._
import javafx.beans.value.ObservableValue
import javafx.collections.{ListChangeListener, FXCollections}
import scala.collection.JavaConverters._
import scala.collection.mutable.ArrayBuffer
import javafx.collections.ListChangeListener.Change
import java.util.function.Predicate

object CRUD extends JFXApp {
  val prefix = new TextField() { prefWidth = 60 }
  val name = new TextField() { prefWidth = 100 }
  val surname = new TextField() { prefWidth = 100 }
  val create = new Button("Create")
  val update = new Button("Update") { disable = true }
  val delete = new Button("Delete") { disable = true }
  val entries = new ListView[String]()
  entries.getSelectionModel.selectionMode = SelectionMode.SINGLE

  val externDb = ArrayBuffer[String]("Emil, Hans", "Mustermann, Max", "Tisch, Roman")
  val db = FXCollections.observableArrayList(externDb.asJava)
  db.addListener(new ListChangeListener[String] {
    override def onChanged(c: Change[_ <: String]): Unit = {
      while (c.next) {
        if (c.wasReplaced()) externDb.update(c.getFrom, c.getAddedSubList.get(0))
        else {
          if (c.wasAdded()) externDb.append(c.getAddedSubList.get(0))
          if (c.wasRemoved()) externDb.remove(c.getFrom)
        }
      }
    }
  })
  val dbView = db.filtered(new Predicate[String] {
    override def or(other: Predicate[_ >: String]): Predicate[String] = ???
    override def and(other: Predicate[_ >: String]): Predicate[String] = ???
    override def negate(): Predicate[String] = ???
    override def test(t: String): Boolean = true
  })
  entries.setItems(dbView)

  // The following doesn't work: val fullname = surname.text + ", " + name.text
  val fullname = surname.textProperty.concat(", ").concat(name.textProperty)
  val selectedIndex = entries.getSelectionModel.selectedIndexProperty()
  prefix.text.addListener((v: ObservableValue[_ <: String], o: String, n: String) =>
    dbView.setPredicate(new Predicate[String] {
      override def or(other: Predicate[_ >: String]): Predicate[String] = ???
      override def and(other: Predicate[_ >: String]): Predicate[String] = ???
      override def negate(): Predicate[String] = ???
      override def test(t: String): Boolean = t.startsWith(n)
    }))
  create.onAction = (event: ActionEvent) => db.add(fullname.get)
  delete.onAction = (event: ActionEvent) => db.remove(dbView.getSourceIndex(selectedIndex.get))
  update.onAction = (event: ActionEvent) => db.set(dbView.getSourceIndex(selectedIndex.get), fullname.get)
  delete.disable <== selectedIndex === -1
  update.disable <== selectedIndex === -1

  stage = new PrimaryStage {
    title = "CRUD"
    scene = new Scene {
      content = new BorderPane() {
        padding = Insets(10)
        prefWidth = 400
        prefHeight = 400
        top = new HBox(10) {
          padding = Insets(0,0,10,0)
          alignment = Pos.BASELINE_LEFT
          content = Seq(new Label("Filter prefix: "), prefix)
        }
        center = entries
        right = new GridPane() {
          padding = Insets(0,0,0,10)
          hgap = 10
          vgap = 10
          addRow(0, new Label("Name: "), name)
          addRow(1, new Label("Surname: "), surname)
        }
        bottom = new HBox(10) {
          padding = Insets(10,0,0,0)
          content = Seq(create, update, delete)
        }
      }
    }
  }
}

